File: packages/core/src/storage/schema.ts
import type { OpEntity, SessionSegmentType } from "../domain";
import { SqliteDb } from "./db";

export interface RepoTable {
  id: string;
  name: string;
  color: string | null;
  user_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface StreamTable {
  id: string;
  repo_id: string;
  name: string;
  visibility: "personal" | "shared";
  user_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface IssueTable {
  id: string;
  stream_id: string;
  title: string;
  status: "todo" | "active" | "done" | "abandoned";
  estimate_minutes: number | null;
  notes: string | null;
  todo_for_date: string | null;
  user_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface SessionTable {
  id: string;
  issue_id: string;
  start_time: string;
  end_time: string | null;
  duration_seconds: number | null;
  notes: string | null;
  user_id: string;
  device_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface StashTable {
  id: string;
  repo_id: string | null;
  stream_id: string | null;
  issue_id: string | null;
  session_id: string | null;
  segment_type: SessionSegmentType | null;
  segment_started_at: string | null;
  elapsed_seconds: number | null;
  note: string | null;
  user_id: string;
  device_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface OpTable {
  id: string;
  user_id: string;
  device_id: string;
  entity: OpEntity;
  entity_id: string;
  action: "create" | "update" | "delete" | "restore";
  payload: string;
  timestamp: string;
}

export interface CoreSettingsTable {
  user_id: string;
  device_id: string;
  timer_mode: "stopwatch" | "structured";
  breaks_enabled: number;
  work_duration_minutes: number;
  short_break_minutes: number;
  long_break_minutes: number;
  long_break_enabled: number;
  cycles_before_long_break: number;
  auto_start_breaks: number;
  auto_start_work: number;
  created_at: string;
  updated_at: string;
}

export interface SessionSegmentsTable {
  id: string;
  user_id: string;
  device_id: string;
  session_id: string;
  segment_type: SessionSegmentType;
  elapsed_offset_seconds: number | null;
  start_time: string;
  end_time: string | null;
  created_at: string;
}

export interface ActiveContextTable {
  user_id: string;
  device_id: string;
  repo_id: string | null;
  stream_id: string | null;
  issue_id: string | null;
  updated_at: string;
}

export interface ScratchPadMetaTable {
  id: string;
  user_id: string;
  device_id: string;
  name: string;
  path: string;
  last_opened_at: string;
  pinned: number;
}

export interface DB {
  repos: RepoTable;
  streams: StreamTable;
  issues: IssueTable;
  sessions: SessionTable;
  stash: StashTable;
  ops: OpTable;
  core_settings: CoreSettingsTable;
  session_segments: SessionSegmentsTable;
  active_context: ActiveContextTable;
  scratch_pad_meta: ScratchPadMetaTable;
}

export async function initSchema(): Promise<void> {
  const db = SqliteDb.getDB();
  await db.schema
    .createTable("repos")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("name", "text", (col) => col.notNull().unique())
    .addColumn("color", "text")
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("created_at", "text", (col) => col.notNull())
    .addColumn("updated_at", "text", (col) => col.notNull())
    .addColumn("deleted_at", "text")
    .execute();

  await db.schema
    .createTable("streams")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("repo_id", "text", (col) => col.notNull())
    .addColumn("name", "text", (col) => col.notNull())
    .addColumn("visibility", "text", (col) => col.notNull())
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("created_at", "text", (col) => col.notNull())
    .addColumn("updated_at", "text", (col) => col.notNull())
    .addColumn("deleted_at", "text")
    .execute();

  await db.schema
    .createTable("issues")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("stream_id", "text", (col) => col.notNull())
    .addColumn("title", "text", (col) => col.notNull())
    .addColumn("status", "text", (col) => col.notNull())
    .addColumn("estimate_minutes", "integer")
    .addColumn("notes", "text")
    .addColumn("todo_for_date", "text")
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("created_at", "text", (col) => col.notNull())
    .addColumn("updated_at", "text", (col) => col.notNull())
    .addColumn("deleted_at", "text")
    .execute();

  await db.schema
    .createTable("sessions")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("issue_id", "text", (col) => col.notNull())
    .addColumn("start_time", "text", (col) => col.notNull())
    .addColumn("end_time", "text")
    .addColumn("duration_seconds", "integer")
    .addColumn("notes", "text")
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("device_id", "text", (col) => col.notNull())
    .addColumn("created_at", "text", (col) => col.notNull())
    .addColumn("updated_at", "text", (col) => col.notNull())
    .addColumn("deleted_at", "text")
    .execute();

  await db.schema
    .createTable("stash")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("repo_id", "text")
    .addColumn("stream_id", "text")
    .addColumn("issue_id", "text")
    .addColumn("session_id", "text")
    .addColumn("segment_type", "text")
    .addColumn("segment_started_at", "text")
    .addColumn("elapsed_seconds", "integer")
    .addColumn("note", "text")
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("device_id", "text", (col) => col.notNull())
    .addColumn("created_at", "text", (col) => col.notNull())
    .addColumn("updated_at", "text", (col) => col.notNull())
    .addColumn("deleted_at", "text")
    .execute();

  await db.schema
    .createTable("ops")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("device_id", "text", (col) => col.notNull())
    .addColumn("entity", "text", (col) => col.notNull())
    .addColumn("entity_id", "text", (col) => col.notNull())
    .addColumn("action", "text", (col) => col.notNull())
    .addColumn("payload", "text", (col) => col.notNull())
    .addColumn("timestamp", "text", (col) => col.notNull())
    .execute();

  await db.schema
    .createTable("core_settings")
    .ifNotExists()
    .addColumn("user_id", "text", (col) => col.primaryKey())
    .addColumn("device_id", "text", (col) => col.notNull())
    .addColumn("timer_mode", "text", (col) => col.notNull())
    .addColumn("breaks_enabled", "integer", (col) => col.notNull())
    .addColumn("work_duration_minutes", "integer", (col) => col.notNull())
    .addColumn("short_break_minutes", "integer", (col) => col.notNull())
    .addColumn("long_break_minutes", "integer", (col) => col.notNull())
    .addColumn("long_break_enabled", "integer", (col) => col.notNull())
    .addColumn("cycles_before_long_break", "integer", (col) => col.notNull())
    .addColumn("auto_start_breaks", "integer", (col) => col.notNull())
    .addColumn("auto_start_work", "integer", (col) => col.notNull())
    .addColumn("created_at", "text", (col) => col.notNull())
    .addColumn("updated_at", "text", (col) => col.notNull())
    .execute();

  await db.schema
    .createTable("session_segments")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("device_id", "text", (col) => col.notNull())
    .addColumn("session_id", "text", (col) => col.notNull())
    .addColumn("segment_type", "text", (col) => col.notNull())
    .addColumn("elapsed_offset_seconds", "integer")
    .addColumn("start_time", "text", (col) => col.notNull())
    .addColumn("end_time", "text")
    .addColumn("created_at", "text", (col) => col.notNull())
    .execute();

  await db.schema
    .createTable("active_context")
    .ifNotExists()
    .addColumn("user_id", "text", (col) => col.primaryKey())
    .addColumn("device_id", "text", (col) => col.notNull())
    .addColumn("repo_id", "text")
    .addColumn("stream_id", "text")
    .addColumn("issue_id", "text")
    .addColumn("updated_at", "text", (col) => col.notNull())
    .execute();

  await db.schema
    .createTable("scratch_pad_meta")
    .ifNotExists()
    .addColumn("id", "text", (col) => col.primaryKey())
    .addColumn("user_id", "text", (col) => col.notNull())
    .addColumn("device_id", "text", (col) => col.notNull())
    .addColumn("name", "text", (col) => col.notNull())
    .addColumn("path", "text", (col) => col.notNull().unique())
    .addColumn("last_opened_at", "text", (col) => col.notNull())
    .addColumn("pinned", "integer", (col) => col.notNull())
    .execute();


  // indexes
  await db.schema
    .createIndex("idx_streams_repo_id")
    .on("streams")
    .column("repo_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_issues_stream_id")
    .on("issues")
    .column("stream_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_sessions_issue_id")
    .on("sessions")
    .column("issue_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_stash_repo_id")
    .on("stash")
    .column("repo_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_stash_stream_id")
    .on("stash")
    .column("stream_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_stash_issue_id")
    .on("stash")
    .column("issue_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_ops_entity_entity_id")
    .on("ops")
    .columns(["entity", "entity_id"])
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_repos_user_id")
    .on("repos")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_streams_user_id")
    .on("streams")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_issues_user_id")
    .on("issues")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_sessions_user_id")
    .on("sessions")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_stash_user_id")
    .on("stash")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_ops_user_id")
    .on("ops")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_session_segments_session_id")
    .on("session_segments")
    .column("session_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_session_segments_user_id")
    .on("session_segments")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_active_context_user_id")
    .on("active_context")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_active_context_device_id")
    .on("active_context")
    .column("device_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_scratch_pad_meta_user_id")
    .on("scratch_pad_meta")
    .column("user_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_scratch_pad_meta_device_id")
    .on("scratch_pad_meta")
    .column("device_id")
    .ifNotExists()
    .execute();

  await db.schema
    .createIndex("idx_scratch_pad_meta_last_opened_at")
    .on("scratch_pad_meta")
    .column("last_opened_at")
    .ifNotExists()
    .execute();
}

File: kernel/internal/store/schema.go
package store

import (
	"context"
	"database/sql"
	"fmt"

	sharedtypes "crona/shared/types"

	"github.com/uptrace/bun"
)

func InitSchema(ctx context.Context, db *bun.DB) error {
	models := []any{
		(*RepoModel)(nil),
		(*StreamModel)(nil),
		(*IssueModel)(nil),
		(*HabitModel)(nil),
		(*HabitCompletionModel)(nil),
		(*SessionModel)(nil),
		(*StashModel)(nil),
		(*OpModel)(nil),
		(*CoreSettingsModel)(nil),
		(*SessionSegmentModel)(nil),
		(*ActiveContextModel)(nil),
		(*ScratchPadMetaModel)(nil),
		(*DailyCheckInModel)(nil),
	}

	for _, model := range models {
		if _, err := db.NewCreateTable().Model(model).IfNotExists().Exec(ctx); err != nil {
			return err
		}
	}

	if err := ensureIssueColumn(ctx, db, "completed_at"); err != nil {
		return err
	}
	if err := ensureIssueColumn(ctx, db, "abandoned_at"); err != nil {
		return err
	}
	for _, spec := range []struct {
		table  string
		column string
	}{
		{table: "repos", column: "description"},
		{table: "streams", column: "description"},
		{table: "issues", column: "description"},
	} {
		if err := ensureTextColumn(ctx, db, spec.table, spec.column); err != nil {
			return err
		}
	}
	for columnName, defaultValue := range map[string]string{
		"repo_sort":   "chronological_asc",
		"stream_sort": "chronological_asc",
		"issue_sort":  "priority",
	} {
		if err := ensureCoreSettingsColumn(ctx, db, columnName, defaultValue); err != nil {
			return err
		}
	}
	for columnName, defaultValue := range map[string]int{
		"boundary_notifications_enabled": 1,
		"boundary_sound_enabled":         1,
	} {
		if err := ensureCoreSettingsBoolColumn(ctx, db, columnName, defaultValue); err != nil {
			return err
		}
	}
	if err := ensureHabitCompletionStatusColumn(ctx, db); err != nil {
		return err
	}
	for _, table := range []string{"repos", "streams", "issues", "habits", "habit_completions"} {
		if err := ensurePublicIDColumn(ctx, db, table); err != nil {
			return err
		}
		if err := backfillPublicIDs(ctx, db, table); err != nil {
			return err
		}
	}
	if err := migrateLegacyIssueStatuses(ctx, db); err != nil {
		return err
	}

	indexes := []string{
		"CREATE UNIQUE INDEX IF NOT EXISTS idx_repos_public_id ON repos (public_id)",
		"CREATE UNIQUE INDEX IF NOT EXISTS idx_streams_public_id ON streams (public_id)",
		"CREATE UNIQUE INDEX IF NOT EXISTS idx_issues_public_id ON issues (public_id)",
		"CREATE UNIQUE INDEX IF NOT EXISTS idx_habits_public_id ON habits (public_id)",
		"CREATE UNIQUE INDEX IF NOT EXISTS idx_habit_completions_public_id ON habit_completions (public_id)",
		"CREATE INDEX IF NOT EXISTS idx_streams_repo_id ON streams (repo_id)",
		"CREATE INDEX IF NOT EXISTS idx_issues_stream_id ON issues (stream_id)",
		"CREATE INDEX IF NOT EXISTS idx_habits_stream_id ON habits (stream_id)",
		"CREATE INDEX IF NOT EXISTS idx_habit_completions_habit_id ON habit_completions (habit_id)",
		"CREATE UNIQUE INDEX IF NOT EXISTS idx_habit_completions_habit_date ON habit_completions (habit_id, date, user_id) WHERE deleted_at IS NULL",
		"CREATE INDEX IF NOT EXISTS idx_sessions_issue_id ON sessions (issue_id)",
		"CREATE INDEX IF NOT EXISTS idx_stash_repo_id ON stash (repo_id)",
		"CREATE INDEX IF NOT EXISTS idx_stash_stream_id ON stash (stream_id)",
		"CREATE INDEX IF NOT EXISTS idx_stash_issue_id ON stash (issue_id)",
		"CREATE INDEX IF NOT EXISTS idx_ops_entity_entity_id ON ops (entity, entity_id)",
		"CREATE INDEX IF NOT EXISTS idx_repos_user_id ON repos (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_streams_user_id ON streams (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_issues_user_id ON issues (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_habits_user_id ON habits (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_habit_completions_user_id ON habit_completions (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_sessions_user_id ON sessions (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_stash_user_id ON stash (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_ops_user_id ON ops (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_session_segments_session_id ON session_segments (session_id)",
		"CREATE INDEX IF NOT EXISTS idx_session_segments_user_id ON session_segments (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_active_context_user_id ON active_context (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_active_context_device_id ON active_context (device_id)",
		"CREATE INDEX IF NOT EXISTS idx_scratch_pad_meta_user_id ON scratch_pad_meta (user_id)",
		"CREATE INDEX IF NOT EXISTS idx_scratch_pad_meta_device_id ON scratch_pad_meta (device_id)",
		"CREATE INDEX IF NOT EXISTS idx_scratch_pad_meta_last_opened_at ON scratch_pad_meta (last_opened_at)",
		"CREATE INDEX IF NOT EXISTS idx_daily_checkins_device_id ON daily_checkins (device_id)",
		"CREATE INDEX IF NOT EXISTS idx_daily_checkins_updated_at ON daily_checkins (updated_at)",
	}

	for _, stmt := range indexes {
		if _, err := db.ExecContext(ctx, stmt); err != nil {
			return err
		}
	}

	return nil
}

func migrateLegacyIssueStatuses(ctx context.Context, db *bun.DB) error {
	replacements := map[string]string{
		"todo":   string(sharedtypes.IssueStatusBacklog),
		"active": string(sharedtypes.IssueStatusInProgress),
	}
	for from, to := range replacements {
		if _, err := db.ExecContext(ctx, "UPDATE issues SET status = ? WHERE status = ?", to, from); err != nil {
			return err
		}
	}
	return nil
}

func ensureIssueColumn(ctx context.Context, db *bun.DB, columnName string) error {
	return ensureTextColumn(ctx, db, "issues", columnName)
}

func ensureTextColumn(ctx context.Context, db *bun.DB, tableName string, columnName string) error {
	rows, err := db.QueryContext(ctx, fmt.Sprintf("PRAGMA table_info('%s')", tableName))
	if err != nil {
		return err
	}
	defer func() {
		_ = rows.Close()
	}()

	var (
		cid       int
		name      string
		typ       string
		notnull   int
		dfltValue sql.NullString
		pk        int
	)
	for rows.Next() {
		if err := rows.Scan(&cid, &name, &typ, &notnull, &dfltValue, &pk); err != nil {
			return err
		}
		if name == columnName {
			return nil
		}
	}
	if err := rows.Err(); err != nil {
		return err
	}

	_, err = db.ExecContext(ctx, fmt.Sprintf("ALTER TABLE %s ADD COLUMN %s text", tableName, columnName))
	return err
}

func ensureCoreSettingsColumn(ctx context.Context, db *bun.DB, columnName string, defaultValue string) error {
	rows, err := db.QueryContext(ctx, "PRAGMA table_info('core_settings')")
	if err != nil {
		return err
	}
	defer func() {
		_ = rows.Close()
	}()

	var (
		cid       int
		name      string
		typ       string
		notnull   int
		dfltValue sql.NullString
		pk        int
	)
	for rows.Next() {
		if err := rows.Scan(&cid, &name, &typ, &notnull, &dfltValue, &pk); err != nil {
			return err
		}
		if name == columnName {
			return nil
		}
	}
	if err := rows.Err(); err != nil {
		return err
	}

	_, err = db.ExecContext(ctx, fmt.Sprintf("ALTER TABLE core_settings ADD COLUMN %s text NOT NULL DEFAULT '%s'", columnName, defaultValue))
	return err
}

func ensureCoreSettingsBoolColumn(ctx context.Context, db *bun.DB, columnName string, defaultValue int) error {
	rows, err := db.QueryContext(ctx, "PRAGMA table_info('core_settings')")
	if err != nil {
		return err
	}
	defer func() {
		_ = rows.Close()
	}()

	var (
		cid       int
		name      string
		typ       string
		notnull   int
		dfltValue sql.NullString
		pk        int
	)
	for rows.Next() {
		if err := rows.Scan(&cid, &name, &typ, &notnull, &dfltValue, &pk); err != nil {
			return err
		}
		if name == columnName {
			return nil
		}
	}
	if err := rows.Err(); err != nil {
		return err
	}

	_, err = db.ExecContext(ctx, fmt.Sprintf("ALTER TABLE core_settings ADD COLUMN %s integer NOT NULL DEFAULT %d", columnName, defaultValue))
	return err
}

func ensureHabitCompletionStatusColumn(ctx context.Context, db *bun.DB) error {
	rows, err := db.QueryContext(ctx, "PRAGMA table_info('habit_completions')")
	if err != nil {
		return err
	}

	var (
		cid       int
		name      string
		typ       string
		notnull   int
		dfltValue sql.NullString
		pk        int
	)
	found := false
	for rows.Next() {
		if err := rows.Scan(&cid, &name, &typ, &notnull, &dfltValue, &pk); err != nil {
			_ = rows.Close()
			return err
		}
		if name == "status" {
			found = true
		}
	}
	if err := rows.Err(); err != nil {
		_ = rows.Close()
		return err
	}
	if err := rows.Close(); err != nil {
		return err
	}

	if found {
		_, err := db.ExecContext(ctx, "UPDATE habit_completions SET status = 'completed' WHERE status IS NULL OR status = ''")
		return err
	}

	if _, err := db.ExecContext(ctx, "ALTER TABLE habit_completions ADD COLUMN status text NOT NULL DEFAULT 'completed'"); err != nil {
		return err
	}
	_, err = db.ExecContext(ctx, "UPDATE habit_completions SET status = 'completed' WHERE status IS NULL OR status = ''")
	return err
}
